klib/tun.c

```diff
@@ -0,0 +1,344 @@
+    void (*file_release)(file f);
+} kfuncs;
+
+static err_t tun_if_output(struct netif *netif, struct pbuf *p, const ip4_addr_t *ipaddr)
+{
+    tun t = netif->state;
+    if (kfuncs.enqueue(t->pq, p)) {
+        kfuncs.pbuf_ref(p);
+        return ERR_OK;
+    } else {
+        return ERR_WOULDBLOCK;
+    }
+}
+
+static err_t tun_if_init(struct netif *netif)
+{
+    netif->output = tun_if_output;
+    netif->mtu = 32 * KB;
+    netif->flags = NETIF_FLAG_IGMP | NETIF_FLAG_LINK_UP;
+    return ERR_OK;
+}
+
+closure_function(5, 1, sysreturn, tun_read_bh,
+                 tun, tun, void *, dest, u64, len, thread, t, io_completion, completion,
+                 u64, flags)
+{
+    tun tun = bound(tun);
+    sysreturn ret;
+    struct pbuf *p = kfuncs.dequeue(tun->pq);
+    if (p == INVALID_ADDRESS) {
+        if (tun->f->f.flags & O_NONBLOCK) {
+            ret = -EAGAIN;
+            goto out;
+        }
+        return BLOCKQ_BLOCK_REQUIRED;
+    }
+    void * dest = bound(dest);
+    u64 len = bound(len);
+    if (!(tun->flags & IFF_NO_PI)) {
+        struct tun_pi pi;
+        if (len < sizeof(pi)) {
+            ret = -EINVAL;
+            kfuncs.pbuf_free(p);
+            goto out;
+        }
+        if (len < p->tot_len + sizeof(pi))
+            pi.flags = TUN_PKT_STRIP;
+        else
+            pi.flags = 0;
+        int ip_version = IPH_V((struct ip_hdr *)p->payload);
+        switch (ip_version) {
+        case 4:
+            pi.proto = htons(ETHTYPE_IP);
+            break;
+        case 6:
+            pi.proto = htons(ETHTYPE_IPV6);
+            break;
+        }
+        kfuncs.runtime_memcpy(dest, &pi, sizeof(pi));
+        dest += sizeof(pi);
+        len -= sizeof(pi);
+    }
+    ret = MIN(len, p->tot_len);
+    kfuncs.pbuf_copy_partial(p, dest, ret, 0);
+    kfuncs.pbuf_free(p);
+    if (!(tun->flags & IFF_NO_PI))
+        ret += sizeof(struct tun_pi);
+  out:
+    kfuncs.blockq_handle_completion(tun->bq, flags, bound(completion), bound(t), ret);
+    closure_finish();
+    return ret;
+}
+
+closure_function(1, 6, sysreturn, tun_read,
+                 tun, tun,
+                 void *, dest, u64, len, u64, offset_arg, thread, t, boolean, bh, io_completion, completion)
+{
+    tun tun = bound(tun);
+    if (!tun->netif)
+        return io_complete(completion, t, -EBADFD);
+    blockq_action ba = closure(tun_heap, tun_read_bh, bound(tun), dest, len, t, completion);
+    if (ba == INVALID_ADDRESS)
+        return io_complete(completion, t, -ENOMEM);
+    return kfuncs.blockq_check(tun->bq, t, ba, bh);
+}
+
+closure_function(1, 6, sysreturn, tun_write,
+                 tun, tun,
+                 void *, src, u64, len, u64, offset, thread, t, boolean, bh, io_completion, completion)
+{
+    tun tun = bound(tun);
+    if (!tun->netif)
+        return io_complete(completion, t, -EBADFD);
+    if (!(tun->flags & IFF_NO_PI)) {
+        if (len < sizeof(struct tun_pi))
+            return io_complete(completion, t, -EINVAL);
```

Comment:
Should len == sizeof(struct tun_pi) just return io_complete(completion, t, len) here?

Reply:
Indeed, calling pbuf_alloc() with len=0 is wrong. Fixed.
If the IFF_NO_PI flag is not present, a write() call with len == sizeof(struct tun_pi) returns sizeof(struct tun_pi), while if the IFF_NO_PI flag is present, a write() call with len == 0 returns -EINVAL. This matches the behavior of the Linux kernel.